src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp

```diff
@@ -0,0 +1,269 @@
+      if ( startVertexCost > travelCost )
+      {
+        // vertex is too expensive, discard
+        continue;
+      }
+
+      vertices.insert( j );
+      startPoint = graph->vertex( j ).point();
+
+      // find all edges coming from this vertex
+      for ( int edgeId : graph->vertex( j ).outgoingEdges() )
+      {
+        edge = graph->edge( edgeId );
+        endVertexCost = startVertexCost + edge.cost( 0 ).toDouble();
+        endPoint = graph->vertex( edge.toVertex() ).point();
+        if ( endVertexCost <= travelCost )
+        {
+          // end vertex is cheap enough to include
+          vertices.insert( edge.toVertex() );
+          lines.push_back( QgsPolylineXY() << startPoint << endPoint );
+        }
+        else
+        {
+          // travelCost sits somewhere on this edge, interpolate position
+          QgsPointXY interpolatedEndPoint = QgsGeometryUtils::interpolatePointOnLineByValue( startPoint.x(), startPoint.y(), startVertexCost,
+                                            endPoint.x(), endPoint.y(), endVertexCost, travelCost );
+
+          areaPoints.push_back( interpolatedEndPoint );
+          lines.push_back( QgsPolylineXY() << startPoint << interpolatedEndPoint );
+        }
+      } // edges
+    } // costs
+
+    // convert to list and sort to maintain same order of points between algorithm runs
+    QList< int > verticesList = vertices.toList();
+    std::sort( verticesList.begin(), verticesList.end() );
+    for ( int v : verticesList )
+    {
+      areaPoints.push_back( graph->vertex( v ).point() );
+    }
+
+    if ( pointsSink )
+    {
+      QgsGeometry geomPoints = QgsGeometry::fromMultiPointXY( areaPoints );
+      feat.setGeometry( geomPoints );
+      attributes = sourceAttributes.value( i + 1 );
+      attributes << QStringLiteral( "within" ) << origPoint;
+      feat.setAttributes( attributes );
+      pointsSink->addFeature( feat, QgsFeatureSink::FastInsert );
+
+      if ( includeBounds )
+      {
+        QgsMultiPointXY upperBoundary, lowerBoundary;
+        QVector< int > nodes;
+
+        int vertexId;
+        for ( int v = 0; v < costs.size(); v++ )
+        {
+          if ( costs.at( v ) > travelCost and tree.at( v ) != -1 )
+          {
+            vertexId = graph->edge( tree.at( v ) ).fromVertex();
+            if ( costs.at( vertexId ) <= travelCost )
+            {
+              nodes.push_back( v );
+            }
+          }
+        } // costs
+
+        for ( int n : nodes )
+        {
+          upperBoundary.push_back( graph->vertex( graph->edge( tree.at( n ) ).toVertex() ).point() );
+          lowerBoundary.push_back( graph->vertex( graph->edge( tree.at( n ) ).fromVertex() ).point() );
+        } // nodes
+
+        QgsGeometry geomUpper = QgsGeometry::fromMultiPointXY( upperBoundary );
+        QgsGeometry geomLower = QgsGeometry::fromMultiPointXY( lowerBoundary );
+
+        feat.setGeometry( geomUpper );
+        attributes = sourceAttributes.value( i + 1 );
```

Comment:
The python version didn't have +1 here -- is that intended?

Reply:
My bad. Fixed.